proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,745 @@
+[`dlopen`]/[`dlsym`] interface could be approximated:
+```wasm
+(instance
+  (export "dlopen" (func
+    (param $module moduleref)
+    (param $imports (array (record (field "name" string) (field "value" anyref))))
+    (result instanceref)
+  ))
+  (export "dlsym" (func
+    (param $instance instanceref)
+    (param $symbol string)
+    (result funcref)
+  ))
+)
+```
+
+Thus, as a conservative first step toward runtime dynamic linking, this proposal
+includes `ref.module` and `ref.instance` in the general wasm instruction set and
+includes module and instance reference types in the general [`reftype`] set.
+
+Lastly, since modules can not just import individual functions, but also individual
+memories, tables and globals, new `ref.memory`, `ref.table` and `ref.global`
+instructions and reference types are added:
+```wasm
+ref.memory $memory : [] -> [ (ref (memory ...)) ]
+ref.table $table : [] -> [ (ref (table ...)) ]
+ref.global $global : [] -> [ (ref (global ...)) ]
+```
+(These instructions have been planned since the first Reference Types proposal.)
+
+
+### Nested Modules
+
+Symmetric to nested instances, modules can contain *nested modules* via
+`module` definitions. Nested modules are injected into the same module index
+space as module imports and thus can be instantiated the same way. For example:
+```wasm
+(module
+  (module $child
+    (func $hi (export "hi")
+      ...
+    )
+  )
+  (instance $i (instance.instantiate $child))
+  (func (export "run")
+    call $i.$hi
+  )
+)
+```
+
+Unlike most source-language nested functions/classes, nested modules have no
+special access to their parents' state. However, since modules and types are
+closed, stateless expressions and thus "sharing" is equivalent to duplication,
+nested modules can refer to their parents' types and modules via *aliases*. As
+syntactic sugar in the textual format, a nested module can simply use the
+`$name` of the enclosing module/type:
+```wasm
+(module
+  (type $WasiFile (instance $wasi-file
+    (export "read" (func (param i32 i32 i32) (result i32)))
+  ))
+  (module $child
+    (import "wasi_file" (instance (type $WasiFile)))
+  )
+)
+```
+The desugared form of `$child` would explicitly create the alias with a `type`
+definition which references the parent:
+```wasm
+(module $Parent
+  (type $WasiFile (instance $wasi-file
+    (export "read" (func (param i32 i32 i32) (result i32)))
+  ))
+  (module $child
+    (type $WasiFile (alias $Parent.$WasiFile))
+    (import "wasi_file" (instance (type $WasiFile)))
+  )
+)
+```
+In the binary format, a type alias could be represented by a [De Bruijn index]
+relative to the sequence of parent modules and their respective module/type
+index spaces.
+
+Other than type/module sharing, which can easily be undone by duplication, a
+module A that imports module B is semantically equivalent to module A nesting
+module B. This means tools can merge multiple `.wasm` files into one `.wasm` by
```

Comment:
I think I know what you mean, but this probably needs rewording. Obviously, these to things are different (one requires clients to explicitly provide B to A, the other does not). It's more like that they can "evaluate" to equivalent instance "values".

Reply:
Ok, softened the wording a bit not to make too strong of a claim.